fix(multi-party-conference): implement-multi-party-conference-feature

akulakum opened the pull request:
<!--
Hey there,\
Thank you for taking the time to improve our code! 🙂\
Please let us know why this change is necessary and what testing you have performed. \
This ensures our reviewers understand the impact of your change. \

**IMPORTANT**
FAILING TO FILL OUT THIS TEMPLATE WILL RESULT IN REJECTION OF YOUR PULL REQUEST
This is for compliance purposes with FedRAMP program.
-->

# COMPLETES #https://jira-eng-sjc12.cisco.com/jira/browse/CAI-7182

Vidcast Link for the MPC Feature: https://app.vidcast.io/share/c3d5e394-c931-45eb-8d76-10e1f1994b33
Transfer Conference Feature: https://app.vidcast.io/share/daf96a54-dbda-419c-9150-e9ba7aea4ca3
Vidcast showcasing flow fixes:  https://app.vidcast.io/share/4a05ae76-c091-4ce1-8bed-dd4309fcb1f3

## This pull request addresses

Implemented the **Multi-Party Conference feature** supporting more than three participants, along with the **Conference Transfer** functionality.

Conference Transfer Behavior:

An agent can initiate a transfer only during a consultation.

**Scenario 1:** If Agent A1 and Agent A2 are on a consultation while the customer is on hold, upon transfer, A1 leaves and A2 takes ownership of the call.

**Scenario 2:** If Agent A1 and Agent A3 are on a consultation while the customer and Agent A2 are already in the main conference, upon transfer, A1 leaves, A2 becomes the new owner, and A3 joins the main conference with A2 and the customer.

## by making the following changes

< DESCRIBE YOUR CHANGES >

<!-- You may include screenshots -->

### Change Type

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

**Consult Conference**

- As an agent, I should be able to add multiple parties to the call via:
      1. Available agent
      2. Available agent in another queue
      3. Dial number

- As an agent, I must be able to add up to 7 participants along with the customer.
- As an agent, I must be able to end the call with the consulted party.
- As the consulting agent, I must be able to leave the conference without ending the original call.
- As an agent, I must be able to transfer the call to the consulted agent as a soft transfer.
- As an agent, I must be able to consult every participant before adding them to the ongoing call.


| **Test Case ID** | **Scenario** | **Steps** | **Expected Result** |
|------------------|--------------|------------|----------------------|
| **TC-01** | Initiate a multi-party conference (more than 3 participants) | 1. Agent A initiates a call with the customer.<br>2. Agent A consults Agent B.<br>3. Agent A adds Agent C to the conference. | All participants (A, B, C, and customer) should join successfully. |
| **TC-02** | Participant leaves the conference | 1. Any participant leaves the call.<br>2. Verify remaining participants stay connected. | Conference continues without disruption. |
| **TC-03** | Agent transfers the conference | 1. Agent A initiates a conference.<br>2. Agent A transfers it to Agent B. | Ownership and audio transfer correctly. |
| **TC-04** | End user disconnects | 1. Customer disconnects. | Remaining agents stay in conference. |
| **TC-05** | Agent ends the conference | 1. Agent A ends the conference. | All participants disconnect; conference terminates. |
| **TC-06** | UI behavior during conference | 1. Observe UI states when adding participants.<br>2. Verify Start/Exit button visibility. | Proper UI state transitions are displayed. |
| **TC-07** | Conference transfer with customer | 1. Agent A initiates a conference with the customer and Agent B.<br>2. Agent A transfers the conference to Agent B. | Agent B should take ownership, and the customer should remain connected seamlessly. |
| **TC-08** | Conference ownership transfer after initiator exit | 1. Customer and Agent1 are connected.<br>2. Agent1 consults and adds Agent2 to the conference.<br>3. Agent1 exits from the conference and transfers the call to Agent2. | Agent1 should be able to leave the interaction and go to wrap-up state.<br>Next oldest agent participant (Agent2) should own the interaction. |
| **TC-09** | Consult and transfer involving multiple agents | 1. Customer, Agent1, and Agent2 are in a conference.<br>2. Agent1 performs a consult with Agent3.<br>3. Agent1 connects to Agent3 and transfers the conference before leaving. | Agent1 should be able to leave the interaction and go to wrap-up state.<br>Agent3 joins the conference with customer and Agent2.<br>Next oldest agent (Agent2) becomes the interaction owner as Agent1 (primary) left. |
| **TC-10** | Consult and transfer with external DN | 1. Customer, Agent1, and External DN are in a conference.<br>2. Agent1 performs any type of consult and connects to Agent3.<br>3. Agent1 transfers the conference and leaves. | Agent1 should be able to leave the interaction and go to wrap-up state.<br>Agent3, now part of the interaction, becomes the new interaction owner. |

</body></html>

## The GAI Coding Policy And Copyright Annotation Best Practices ##
 
<!-- **MANDATORY** If Yes, Mention the GAI Coding Policy Copyright Annotation Best Practices followed separated by a comma below the yes checkbox -->
 
- [ ] GAI was not used (or, no additional notation is required)
- [ ] Code was generated entirely by GAI
- [x] GAI was used to create a draft that was subsequently customized or modified
- [ ] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [ ] Tool used for AI assistance (GitHub Copilot / Other - specify)
  - [ ] Github Copilot
  - [x] Other - Please Specify
- [ ] This PR is related to
  - [x] Feature
  - [ ] Defect fix
  - [ ] Tech Debt
  - [ ] Automation

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request
- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

akulakum committed:
fix(multy-party-conference): implement-multy-party-conference-feature

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

rsarika committed:
fix: fix user defined properties not to delete

rsarika committed:
fix: updated types

rsarika committed:
fix: updated task utils to function approach

rsarika committed:
fix: desktop mode not working

akulakum committed:
fix(multy-party-conference): implement-sample-appjs-changes

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

akulakum committed:
fix(multy-party-conference): fix-taskUpdate-issue

rsarika committed:
feat(conference): enhance task handling and UI updates for consultations

rsarika committed:
fix: added await for login

rsarika committed:
fix: removed ravi

rsarika committed:
feat(call-control): streamline UI updates and control management for consultation states

rsarika committed:
fix: conference transfer

rsarika committed:
fix: cleanup

rsarika committed:
fix: mute element fix for extension

rsarika committed:
Merge branch 'next' of github.com:webex/webex-js-sdk into MULTI_PARTY_CONFERENCE

rsarika committed:
fix: cleanup of isConferenceInProgress

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

rsarika committed:
fix: removed unused function

rsarika committed:
fix: double webex initialization

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -1,5 +1,5 @@
 // Globals
-let webex;
+let webex = undefined;
```
Why do we have to do this?

akulakum replied:
To fix the double initialization and browser call issue

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -46,6 +42,22 @@ const updateDialNumberElm  = document.querySelector('#updateDialNumber');
+  </label>
+  <br/><small style="color:#666;">📝 MPC Testing: Simulates the consultation interaction behavior</small>
+`;
+incomingDetailsElm.parentNode.appendChild(allowInteractElm);
```
It looks like there's no dynamic value involved here. Why don't we put it in HTML and just show/hide the div instead?

akulakum replied:
Moved the logic to the HTML as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -897,20 +970,7 @@ function registerTaskListeners(task) {
-      updateTaskList(); // Update the task list UI to have latest tasks
-      handleTaskSelect(task);
-    }
+    updateTaskList(); // Update the task list UI to have latest tasks
```
nitpick. 

Could have been simplified like this:
```
task.on('task:end', updateTaskList); 
```

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -922,119 +982,53 @@ function registerTaskListeners(task) {
   // Consult flows
   task.on('task:consultCreated', (task) => {
-    console.info('Consult created');
+    updateTaskList();
```
Same as [this comment](https://github.com/webex/webex-js-sdk/pull/4520/files#r2432225172)

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -922,119 +982,53 @@ function registerTaskListeners(task) {
 
   task.on('task:offerConsult', (task) => {
-    console.info('Received consult offer from another agent');
+    updateTaskList();
```
Same as [this comment](https://github.com/webex/webex-js-sdk/pull/4520/files#r2432225172)

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -922,119 +982,53 @@ function registerTaskListeners(task) {
-      showEndConsultButton();
-      consultTransferBtn.disabled = true; // Disable the consult transfer button since we are not yet owner of the call
-    }
+    updateTaskList();
```
Same as [this comment](https://github.com/webex/webex-js-sdk/pull/4520/files#r2432225172)

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -922,119 +982,53 @@ function registerTaskListeners(task) {
-      consultTransferBtn.style.display = 'inline-block'; // Show the consult transfer button
-      consultTransferBtn.disabled = false; // Enable the consult transfer button
-    }
+   updateTaskList();
```
Same as [this comment](https://github.com/webex/webex-js-sdk/pull/4520/files#r2432225172)

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -922,119 +982,53 @@ function registerTaskListeners(task) {
-      hideEndConsultButton();
-      showConsultButton();
-    }
+     updateTaskList();
```
Same as [this comment](https://github.com/webex/webex-js-sdk/pull/4520/files#r2432225172)

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -922,119 +982,53 @@ function registerTaskListeners(task) {
-      enableTransferControls();
-      enableCallControlPostConsult();
-    }
+    updateTaskList();
```
Same as [this comment](https://github.com/webex/webex-js-sdk/pull/4520/files#r2432225172)

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -922,119 +982,53 @@ function registerTaskListeners(task) {
-        task = undefined;
-      }
-    }
+   updateTaskList();
```
Same as [this comment](https://github.com/webex/webex-js-sdk/pull/4520/files#r2432225172)

akulakum replied:
Made the change as suggested.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -1047,36 +1041,141 @@ function disableAllCallControls() {
+  // Helper to set multiple controls at once
+  function setControls(configs) {
+    for (const [elm, [hide, disable]] of Object.entries(configs)) {
+      makeDisabledAndHide(eval(elm), hide, disable);
```
It is not a good practice to use `eval` even though it is in sample app. eval can execute anything and might open up to memory leaks or other vulenrabilities.

akulakum replied:
Addressed

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -1085,48 +1184,109 @@ function updateCallControlUI(task) {
+    const controlMap = {
+      beingConsulted: () => {}, // No changes
+      beingConsultedAccepted: () => setControls({
+        'holdResumeElm': [true, false],
```
We will have hard time debugging in sample app if we do not have these as maps instead of array. This gives more readability:
```
'holdResumeElm': {
   'hide': true,
   'disabled': false
}
```

akulakum replied:
Adressed

mkesavan13 commented on packages/@webex/contact-center/src/services/task/constants.ts:
```diff
@@ -23,6 +23,25 @@ export const CONFERENCE_TRANSFER = '/conference/transfer';
+  WRAP_UP_REQUIRED: 'wrapUpRequired',
+  /** Indicates if a conference is currently in progress (2+ active agents) */
+  IS_CONFERENCE_IN_PROGRESS: 'isConferenceInProgress',
+} as const;
```
Why do we have to say as `const` while the declaration itself is `const`?

akulakum replied:
Removed it. Thanks for pointing that out.

mkesavan13 commented on packages/@webex/contact-center/src/services/task/index.ts:
```diff
@@ -281,9 +281,24 @@ export default class Task extends EventEmitter implements ITask {
   private reconcileData(oldData: TaskData, newData: TaskData): TaskData {
+    // Remove keys from oldData that are not in newData
+    Object.keys(oldData).forEach((key) => {
+      if (!(key in newData) && !KEYS_TO_NOT_DELETE.includes(key as string)) {
+        delete oldData[key];
+      }
+    });
+
+    // Merge or update keys from newData
     Object.keys(newData).forEach((key) => {
-      if (newData[key] && typeof newData[key] === 'object' && !Array.isArray(newData[key])) {
-        oldData[key] = this.reconcileData({...oldData[key]}, newData[key]);
+      if (
+        newData[key] &&
+        typeof newData[key] === 'object' &&
+        !Array.isArray(newData[key]) &&
+        oldData[key] &&
+        typeof oldData[key] === 'object' &&
+        !Array.isArray(oldData[key])
+      ) {
+        this.reconcileData(oldData[key], newData[key]);
```
If the old data would have the key and new data wouldn't, isn't it good enough if we did a simple object destructuring like this?

```
oldData = {
   ...oldData,
   ...newData
}
```

akulakum replied:
Deep Merging: We need recursive merging for nested objects. Destructuring would replace entire nested objects instead of merging them.

Conditional Preservation: We only delete keys that aren't in KEYS_TO_NOT_DELETE. Simple spread can't handle this selective preservation logic.

The key deletion + recursive merging ensures proper deep merge behavior that destructuring alone cannot achieve.

mkesavan13 replied:
Makes sense. There's no native method for deep merge anyway. Resolving the comment

mkesavan13 commented on packages/@webex/contact-center/src/services/task/index.ts:
````diff
@@ -1707,7 +1719,7 @@ export default class Task extends EventEmitter implements ITask {
    * ```
    */
-  /* public async transferConference(): Promise<TaskResponse> {
+  public async transferConference(): Promise<TaskResponse> {
````
Was the transfer conference method already implemented just not used so far? If not, how is this just an uncomment in the changeset?

akulakum replied:
Yes, the transferConference method was implemented earlier during the development of the three-party conference feature. However, since the transfer functionality wasn’t required for that scope, Priya had requested that we comment it out until we added support for multi-party conferences. In this PR, we are extending support for multi-party conferences (more than three participants), so the transfer conference logic is now required and has been uncommented accordingly

mkesavan13 commented on packages/@webex/contact-center/src/services/task/TaskManager.ts:
```diff
@@ -358,18 +363,55 @@ export default class TaskManager extends EventEmitter {
             // Participant joined conference - update task state with participant information and emit event
-            task = this.updateTaskData(task, payload.data);
+            // Pre-calculate isConferenceInProgress with updated data to avoid double update
+            const simulatedTaskForJoin = {
+              ...task,
+              data: {...task.data, ...payload.data},
+            };
+            task = this.updateTaskData(task, {
+              ...payload.data,
+              isConferenceInProgress: getIsConferenceInProgress(simulatedTaskForJoin),
+            });
```
Good one 💚

mkesavan13 commented on packages/@webex/contact-center/src/services/task/TaskManager.ts:
```diff
@@ -358,18 +363,55 @@ export default class TaskManager extends EventEmitter {
+              ...payload.data,
+              isConferenceInProgress: getIsConferenceInProgress(simulatedTaskForLeft),
+            });
+            if (checkParticipantNotInInteraction(task, this.agentId)) {
+              if (
+                isParticipantInMainInteraction(task, this.agentId) ||
```
Aren't these functions negation of each other? 

If `checkParticipantNotInInteraction` is `true`, `isParticipantInMainInteraction` should always be `false`

Is my understanding correct here?

akulakum replied:
The first one is to check whether it is an interaction or not.  The second one is whether it is in main interaction(mediaObj.mType === 'mainCall')

mkesavan13 replied:
Exactly.

So, if the first one says that the participant is not in any interaction, they are definitely not in the main interaction. Right?

akulakum replied:
There is a small difference of condition hasLeft in first method. This logic also we copied from Agent desktop.

mkesavan13 replied:
I see. So, is this like saying, "They were part of the conversation earlier but not now"?

akulakum replied:
Yes, they were part of the conversation, but hasLeft and in the second condition we are checking whether they were part of main interaction or not.

mkesavan13 commented on packages/@webex/contact-center/src/services/task/TaskUtils.ts:
```diff
@@ -0,0 +1,94 @@
+  const {data} = task;
+
+  return (
+    !(agentId in data.interaction.participants) ||
+    (agentId in data.interaction.participants && data.interaction.participants[agentId].hasLeft)
```
Is participant list an array or object with agentId as key? Could you please paste a sample participant list?

akulakum replied:
Yes, it's object with agentId as key.

[participantLeft.txt](https://github.com/user-attachments/files/22945976/participantLeft.txt)

mkesavan13 commented on packages/@webex/contact-center/src/services/task/types.ts:
````diff
@@ -1224,25 +1226,109 @@ export interface ITask extends EventEmitter {
+   * await task.toggleMute();
+   * ```
+   */
+  toggleMute(): Promise<void>;
````
Did we add toggleMute only now?

akulakum replied:
We have added, but we did not provide in ITask, so I have updated ITask with all the missing methods.

mkesavan13 commented on packages/@webex/contact-center/test/unit/spec/services/task/index.ts:
```diff
@@ -335,6 +336,60 @@ describe('Task', () => {
       expect(task.data).toEqual(expectedData);
     });
+
+    it('test updating the task data by merging and preserving existing keys', async () => {
```
```suggestion
    it('updates the task data by merging and preserving existing keys', async () => {
```

akulakum committed:
fix(multy-party-conference): address-review-comments

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

akulakum committed:
fix(multy-party-conference): address-review-comments

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -46,6 +42,13 @@ const updateDialNumberElm  = document.querySelector('#updateDialNumber');
+const participantListElm = document.createElement('div');
+participantListElm.id = 'participant-list';
+participantListElm.style.cssText = 'margin-top:10px;padding:8px;border:1px solid #ccc;background:#f9f9f9;display:none;';
+incomingDetailsElm.parentNode.appendChild(participantListElm);
```
This seems static again. There's no dynamic stuff here. Why not keep it in the HTML itself and just hide and show?

akulakum replied:
Addressed

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -1047,36 +1016,159 @@ function disableAllCallControls() {
+  consultTransferBtn.style.display = 'none';
+}
+
+function makeDisabledAndHide(element, hide, disable)
+{
+  element.style.display = hide ? 'none' : 'inline-block';
+  element.disabled = disable;
+}
+
+function isSecondaryAgent(task) {
+  const interaction = task.data.interaction;
+
+  return (
+    interaction.callProcessingDetails.relationshipType === 'consult' &&
+    interaction.callProcessingDetails.parentInteractionId &&
+    interaction.callProcessingDetails.parentInteractionId !== interaction.interactionId
+  );
+}
+
+function isSecondaryEpDnAgent(task) {
+  return task.data.interaction.mediaType === 'telephony' && isSecondaryAgent(task);
+}
```
What is the difference here?

akulakum replied:
We copied the logic from Agent Desktop and there is an additional condition of telephony and it might be useful when we implement conference feature in digital channel.

mkesavan13 replied:
Can we add short descriptions for each methods to differentiate? Just in case if someone else looks into these in the future, it'll be helpful for them

akulakum replied:
Added the description.

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -1047,36 +1016,159 @@ function disableAllCallControls() {
+  return interaction?.state;
+}
+
+function getStatus(task, agentId) {
```
Should we call it something more specific? `getTaskStatus` or perhaps `getConferenceStatus`?

akulakum replied:
Addressed

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

mkesavan13 commented on docs/samples/contact-center/app.js:
```diff
@@ -1085,48 +1177,109 @@ function updateCallControlUI(task) {
+      }),
+      consultInitiated: () => setControls({
+        'holdResumeElm': { hide: true, disable: false },
+        'muteElm': { hide: true || !isBrowser, disable: false },
```
Wouldn't this `hide` always be `true`?

akulakum replied:
Adressed

mkesavan13 reviewed (commented):
A few more tiny comments and we are good to go

akulakum committed:
fix(multy-party-conference): address-review-comments

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

mkesavan13 reviewed (approved):
I have left two comments unresolved. Kindly respond to those before merging. Approving this PR.

Thanks a lot, Akula and Ravi for working on this

akulakum committed:
fix(multy-party-conference): add-description

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

akulakum committed:
Merge branch 'next' into MULTI_PARTY_CONFERENCE

Kesari3008 commented on packages/@webex/contact-center/src/services/task/TaskManager.ts:
```diff
@@ -358,18 +363,55 @@ export default class TaskManager extends EventEmitter {
             // Conference ended - update task state and emit event
             task = this.updateTaskData(task, payload.data);
-            task.emit(TASK_EVENTS.TASK_CONFERENCE_ENDED, task);
+            if (
```
Won't task manager test file also need to be updated based on these changes

rsarika replied:
addressed

Kesari3008 commented on packages/@webex/contact-center/src/services/task/TaskUtils.ts:
```diff
@@ -0,0 +1,94 @@
+ * @returns true if the agent is the primary agent, false otherwise
+ */
+export const isPrimary = (task: ITask, agentId: string): boolean => {
+  if (!task?.data?.interaction?.owner) {
```
task object seems to be a mandatory param here so why would we need ? after that

rsarika replied:
addressed

Kesari3008 commented on packages/@webex/contact-center/src/services/task/TaskUtils.ts:
```diff
@@ -0,0 +1,94 @@
+ * @param agentId - The agent ID to get status for
+ * @returns Object containing various status flags for the agent
+ */
+export const getParticipantStatus = (task: ITask, agentId: string) => ({
```
Why do we need separate method to set different values if we have methods that directly gives a boolean return value

rsarika replied:
removed this method as we are not using this also.

Kesari3008 commented on packages/@webex/contact-center/src/services/task/TaskUtils.ts:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  return Object.values(task.data.interaction.media).some(
+    (mediaObj: any) => mediaObj.mType === 'mainCall' && mediaObj.participants?.includes(agentId)
```
Please do not use any in the code files. Create a type for it and use that

rsarika replied:
removed

Kesari3008 commented on packages/@webex/contact-center/src/services/task/TaskManager.ts:
```diff
@@ -358,18 +363,55 @@ export default class TaskManager extends EventEmitter {
-            task.emit(TASK_EVENTS.TASK_CONFERENCE_ENDED, task);
+            if (
+              !task ||
+              isPrimary(task, this.agentId) ||
```
Won't the task object also have the agentId, why do we need to pass it separately ?

rsarika replied:
the agentId that comes in task object is not always the logged-in agent's id. eg: participant-left event  will give the agentId who left, so we cannot rely on task data agentId anymore.

Kesari3008 reviewed (changes requested):
Testing Issues/Queries Observed:
1. Without the conference flow if consult is declined or failed, there is no option to resume the call, hold button text says hold and even if I click the button it attempts the hold API

2.  With the conference flow if consult is failed or declined, customer is stuck on hold and there is no way from any participant's view to resume the call for customer.

3. Same behavior is observed for end consult flow.

4. Participant list gets updated with the agent information which is being consulted even before they are added in the conference. Is that expected, please verify ?

5. End call button is enabled at all times for all the participants, there are situations while one of the participants is trying to add another agent and customer is on hold. At this moment neither customer agents won't know what happened if one of the particpants decide to end the conference entirely for everyone. Verify this behavior as well in the agent desktop.

6. If the initiator of the conference exits the conference, partipant list still shows them in participant list but there is cross mark present next to it. Is that expected ?

7. Decline buttons in browser mode is missing sometimes.

8. We should call it merge instead of start conference.

9. Also did we test the conference until 7th participants and what all buttons are enabled for participants at that point in time

rsarika committed:
fix: addressed comments

rsarika committed:
Merge branch 'MULTI_PARTY_CONFERENCE' of github.com:akulakum/webex-js-sdk into MULTI_PARTY_CONFERENCE

akulakum committed:
fix(multy-party-conference): address-review-comments

akulakum committed:
fix(multy-party-conference): fix-resume-issue

akulakum commented:
> Testing Issues/Queries Observed:
> 
> 1. Without the conference flow if consult is declined or failed, there is no option to resume the call, hold button text says hold and even if I click the button it attempts the hold API
> 2. With the conference flow if consult is failed or declined, customer is stuck on hold and there is no way from any participant's view to resume the call for customer.
> 3. Same behavior is observed for end consult flow.
> 4. Participant list gets updated with the agent information which is being consulted even before they are added in the conference. Is that expected, please verify ?
> 5. End call button is enabled at all times for all the participants, there are situations while one of the participants is trying to add another agent and customer is on hold. At this moment neither customer agents won't know what happened if one of the particpants decide to end the conference entirely for everyone. Verify this behavior as well in the agent desktop.
> 6. If the initiator of the conference exits the conference, partipant list still shows them in participant list but there is cross mark present next to it. Is that expected ?
> 7. Decline buttons in browser mode is missing sometimes.
> 8. We should call it merge instead of start conference.
> 9. Also did we test the conference until 7th participants and what all buttons are enabled for participants at that point in time

All the review comments have been addressed, and the corresponding fixes are showcased in the Vidcast.

Vidcast Link: https://app.vidcast.io/share/4a05ae76-c091-4ce1-8bed-dd4309fcb1f3

Kesari3008 commented on packages/@webex/contact-center/src/services/task/TaskUtils.ts:
```diff
@@ -28,7 +28,7 @@ export const isParticipantInMainInteraction = (task: ITask, agentId: string): bo
 
   return Object.values(task.data.interaction.media).some(
-    (mediaObj: any) => mediaObj.mType === 'mainCall' && mediaObj.participants?.includes(agentId)
+    (mediaObj) => mediaObj.mType === 'mainCall' && mediaObj.participants?.includes(agentId)
```
What if mediaObj is undefined or doesn't exist. Since we do not have type for it, we might get run time error for this line : mediaObj.mType

akulakum replied:
Addressed by adding mediaObj && and there is no specific type, but the structure is inferring automatically.  Please find the attached screenshot.
<img width="851" height="346" alt="Screenshot 2025-10-22 at 8 50 13 PM" src="https://github.com/user-attachments/assets/6c3b1674-79db-40ac-8c66-b6289f736b50" />

Kesari3008 reviewed (approved):
1. When you are with 7 participants, conference transfer is something that should still be possible. It's worth checking it and implement it if agent desktop allows it
2. Are we disabling the end button for conference initiator also when any consult is in progress ?

Kesari3008 commented:
Add the vidcast with flow fixes in the description. And update the description with testcases performed manually alon with checking the boxes about GAI and documentaion and coding guidelines.

akulakum commented:
> 1. When you are with 7 participants, conference transfer is something that should still be possible. It's worth checking it and implement it if agent desktop allows it

The consult button itself will be disabled, so only when in consult we can transfer conference or merge, so this is not possible and the behaviour is same in Agent desktop.  If you mean while adding 7th participant, then it's already implemented(we can do transfer conference).   

> 2. Are we disabling the end button for conference initiator also when any consult is in progress ?

Yes we are disabling the end button when consult is in progress where the person who initiates the consult as well as others also who are in conference call.

akulakum committed:
fix(multy-party-conference): addedd-mediaObj-undefined-check

akulakum commented:
> Add the vidcast with flow fixes in the description. And update the description with testcases performed manually alon with checking the boxes about GAI and documentaion and coding guidelines.

Added the Vidcast with flow fixes and updated the PR description with the manually performed test cases. Also ensured the GAI, documentation, and coding guideline checkboxes are updated.
